File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.py
def binary_search(lis, left, right, target):
    while left < right:
        mid = (left + right) // 2
        if lis[mid] < target:
            left = mid + 1
        else:
            right = mid
    return right

def lis_length(arr):
    n = len(arr)
    lis = [0] * n  # LIS를 저장할 배열
    length = 0  # LIS의 길이

    for i in range(n):
        if length == 0 or lis[length - 1] < arr[i]:
            lis[length] = arr[i]
            length += 1
        else:
            pos = binary_search(lis, 0, length, arr[i])  # 직접 구현한 이분 탐색
            lis[pos] = arr[i]

    return length

# 입력 및 실행
n = int(input())
arr = list(map(int, input().split()))
print(lis_length(arr))

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.cc
#include <iostream>
#include <vector>
using namespace std;

int N;
int arr[1000000] = { 0, };
int lis[1000000] = { 0, };

int binary_search(int left, int right, int target) {
	while (left < right) {
		int mid = (left + right) / 2;
		if (lis[mid] < target) {
			left = mid + 1;
		}
		else {
			right = mid;
		}
	}

	return right;
}


int lis_length() {
	int len = 0;
	for (int i = 0;i < N;i++) {
		if (len == 0 || lis[len - 1] < arr[i]) {
			lis[len] = arr[i];
			len++;
		}
		else {
			int position = binary_search(0, len, arr[i]);
			lis[position] = arr[i];
		}
	}

	return len;
}


int main() {
	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> N;

	for (int i = 0;i < N;i++) {
		cin >> arr[i];
	}

	cout << lis_length();

	return 0;
}